core/services/ocr2/plugins/median/plugin.go

```diff
@@ -1,78 +1,243 @@
 
+	"github.com/smartcontractkit/chainlink-relay/pkg/loop"
 	"github.com/smartcontractkit/chainlink-relay/pkg/types"
 
+	v2 "github.com/smartcontractkit/chainlink/v2/core/config/v2"
 	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services"
 	"github.com/smartcontractkit/chainlink/v2/core/services/job"
 	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/median/config"
 	"github.com/smartcontractkit/chainlink/v2/core/services/ocrcommon"
 	"github.com/smartcontractkit/chainlink/v2/core/services/pipeline"
+	"github.com/smartcontractkit/chainlink/v2/core/services/relay"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
 )
 
 type MedianConfig interface {
 	JobPipelineMaxSuccessfulRuns() uint64
+	relay.EnvConfig
 }
 
-// NewMedian parses the arguments and returns a new Median struct.
-func NewMedianServices(jb job.Job,
-	ocr2Provider types.MedianProvider,
+func NewMedianServices(ctx context.Context,
+	jb job.Job,
+	isNewlyCreatedJob bool,
+	relayer loop.Relayer,
 	pipelineRunner pipeline.Runner,
 	runResults chan pipeline.Run,
 	lggr logger.Logger,
-	ocrLogger commontypes.Logger,
 	argsNoPlugin libocr2.OracleArgs,
 	cfg MedianConfig,
 	endpoint commontypes.MonitoringEndpoint,
-) ([]job.ServiceCtx, error) {
+	errorLog loop.ErrorLog,
+) (srvs []job.ServiceCtx, err error) {
 	var pluginConfig config.PluginConfig
-	err := json.Unmarshal(jb.OCR2OracleSpec.PluginConfig.Bytes(), &pluginConfig)
+	err = json.Unmarshal(jb.OCR2OracleSpec.PluginConfig.Bytes(), &pluginConfig)
 	if err != nil {
-		return nil, err
+		return
 	}
 	err = config.ValidatePluginConfig(pluginConfig)
 	if err != nil {
-		return nil, err
+		return
+	}
+	spec := jb.OCR2OracleSpec
+	//TODO retry https://smartcontract-it.atlassian.net/browse/BCF-2112
+	provider, err := relayer.NewMedianProvider(ctx, types.RelayArgs{
+		ExternalJobID: jb.ExternalJobID,
+		JobID:         spec.ID,
+		ContractID:    spec.ContractID,
+		New:           isNewlyCreatedJob,
+		RelayConfig:   spec.RelayConfig.Bytes(),
+	}, types.PluginArgs{
+		TransmitterID: spec.TransmitterID.String,
+		PluginConfig:  spec.PluginConfig.Bytes(),
+	})
+	if err != nil {
+		return
+	}
+	srvs = append(srvs, provider)
+	argsNoPlugin.ContractTransmitter = provider.ContractTransmitter()
+	argsNoPlugin.ContractConfigTracker = provider.ContractConfigTracker()
+	argsNoPlugin.OffchainConfigDigester = provider.OffchainConfigDigester()
+
+	abort := func() {
+		var mc services.MultiClose
+		for i := range srvs {
+			mc = append(mc, srvs[i])
+		}
+		if cerr := mc.Close(); err != nil {
+			lggr.Errorw("Error closing unused services", "err", cerr)
+		}
+	}
+
+	var median loop.PluginMedian
+	if cmdName := v2.EnvMedianPluginCmd.Get(); cmdName != "" {
+		ms := NewPluginMedianService(cmdName, lggr, cfg)
+		if err = ms.Launch(); err != nil {
+			abort()
+			return
+		}
+		median = ms
+		srvs = append(srvs, ms)
+	} else {
+		median = NewPlugin(lggr, ctx.Done())
```

Comment:
do you mean to call ctx.Done() here? if so, why - i've never read code written that way. are trying to pass the func `ctx.Done` ?

Reply:
Passing the returned receive-only channel, since we only care about propagating that signal (and the other caller doesn't have a context).